### ISSUE TITLE
Side effect using xgboost.train with callbacks and subsampling enabled

### ISSUE THREAD

**Issue by ldesreumaux** [dmlc/xgboost#5443]
While running benchmarks using the callback that allows to update the learning rate at each iteration, I noticed a side effect that seems to occur only when row subsampling is enabled.

See the script below to reproduce the issue. In this script, two models are trained using `xgboost.train`: one that does not have callbacks, and one that has a callback that does a "fake learning rate update", which means it performs a call to `set_param` that does not change the learning rate.

Using the same seed, and with early stopping enabled, the two models should build exactly the same number of trees. However, when row subsampling is enabled, the numbers of trees are not the same (the two first print statements at the end of the script give different numbers while the two other print statements give the same numbers).

```
import numpy as np
import xgboost as xgb

def generate_data():
    X = np.random.normal(5, 2, [n_obs, n_features],)
    y = np.random.gamma(2, 4, n_obs)
    return X, y

def fake_update_learning_rate():    
    def callback(env):
        env.model.set_param('learning_rate', learning_rate)    
    return callback

seed = 42
num_boost_round = 1000
early_stopping_rounds = 10
learning_rate = 0.1
tree_method = "hist"

n_obs = 10**6
n_features = 20

params_nosubsample = {
    "learning_rate": learning_rate,
    "tree_method": tree_method,
    "subsample": 1,
    "seed": seed
}

params_withsubsample = {
    "learning_rate": learning_rate,
    "tree_method": tree_method,
    "subsample": 0.9,
    "seed": seed
}

np.random.seed(seed)
X_train, y_train = generate_data()
X_test, y_test = generate_data()
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

model1_nosubsample = xgb.train(
    params=params_nosubsample,
    dtrain=dtrain,
    num_boost_round=num_boost_round,
    early_stopping_rounds=early_stopping_rounds,
    evals=[(dtest, 'test')],
    verbose_eval=True
)

model2_nosubsample = xgb.train(
    params=params_nosubsample,
    dtrain=dtrain,
    num_boost_round=num_boost_round,
    early_stopping_rounds=early_stopping_rounds,
    evals=[(dtest, 'test')],
    verbose_eval=True,
    callbacks=[fake_update_learning_rate()]
)

model1_withsubsample = xgb.train(
    params=params_withsubsample,
    dtrain=dtrain,
    num_boost_round=num_boost_round,
    early_stopping_rounds=early_stopping_rounds,
    evals=[(dtest, 'test')],
    verbose_eval=True
)

model2_withsubsample = xgb.train(
    params=params_withsubsample,
    dtrain=dtrain,
    num_boost_round=num_boost_round,
    early_stopping_rounds=early_stopping_rounds,
    evals=[(dtest, 'test')],
    verbose_eval=True,
    callbacks=[fake_update_learning_rate()]
)

print("With subsampling:", model1_withsubsample.best_iteration, "iterations")
print("With subsampling + callback:", model2_withsubsample.best_iteration, "iterations")
print("Without subsampling:", model1_nosubsample.best_iteration, "iterations")
print("Without subsampling + callback:", model2_nosubsample.best_iteration, "iterations")
```

This issue is probably due to a re-initialization of the seed used in the pseudorandom number generator when `set_param` is called. In the C++ code, `SetParam` sets the flag `need_configuration` to `true`, so when `BoostOneIter` is called, `Configure` is called, and the seed is re-initialized:
```
void Configure() override {
    if (!this->need_configuration_) { return; }
    …

    // set seed only before the model is initialized
    common::GlobalRandom().seed(generic_parameters_.seed);

```

**Comment by hcho3** [comment]
We should document the limitation. The learning-rate update should not be performed if you want deterministic training with a specific random seed.

**Comment by trivialfis** [comment]
This can be fixed by caching the see.  Also will document the `seed_per_iteration`.
